source/common/quic/envoy_quic_utils.cc

```diff
@@ -168,16 +180,17 @@ createConnectionSocket(const Network::Address::InstanceConstSharedPtr& peer_addr
                        Network::Address::InstanceConstSharedPtr& local_addr,
                        const Network::ConnectionSocket::OptionsSharedPtr& options,
                        const bool prefer_gro) {
-  if (local_addr == nullptr) {
-    local_addr = Network::Utility::getLocalAddress(peer_addr->ip()->version());
-  }
+  ASSERT(peer_addr != nullptr);
   size_t max_addresses_cache_size =
       Runtime::runtimeFeatureEnabled(
           "envoy.reloadable_features.quic_upstream_socket_use_address_cache_for_read")
           ? 4u
           : 0u;
   auto connection_socket = std::make_unique<Network::ConnectionSocketImpl>(
-      Network::Socket::Type::Datagram, local_addr, peer_addr,
+      Network::Socket::Type::Datagram,
+      // Use the loopback address if `local_addr` is null, to pass in the socket interface used to
+      // create the IoHandle, without having to make the more expensive `getifaddrs` call.
+      local_addr ? local_addr : getLoopbackAddress(peer_addr->ip()->version()), peer_addr,
```

Comment:
I'm curious how this local address is used. In the server case, we need a local address 'cause that's what we listen. But in the normal POSIX client socket case, a caller typically does not specify a local address. Do you understand how this is used?

Reply:
Unfortunately, the local address is needed so we can get a `SocketInterface` instance because it is needed to create an `IoHandle` instance.  If we don't supply an address here, we segfault in creating the `IoHandle` in the `SocketImpl` constructor.  Dan and I discussed this and we think the setup of SocketImpl isn't correct - we should be able to use a `SocketInterface` independent of an `Address` instance.  But that's a change for a separate PR.

Reply:
I see. That makes sense, but it makes me nervous. Have we tried running this PR against an external server (using HTTP/3)?

Reply:
It should only affect upstream traffic. And if we hide it behind a runtime feature, it should be fine.

Reply:
Excellent. Can you also mention that in the PR description and update the release notes?

Reply:
Release notes was updated in https://github.com/envoyproxy/envoy/pull/35652/files#diff-6f9c718224c533c13c2c0ba1d5abaab86be9d0cc73808749c77934e9f9b0d5d0.  Just updated the PR description.